AutoCollection/ServerRequests.ts

```diff
@@ -0,0 +1,125 @@
+
+    constructor(client:Client) {
+        if (!!AutoCollectServerRequests.INSTANCE) {
+            throw new Error("Server request tracking should be configured from the applicationInsights object");
+        }
+
+        AutoCollectServerRequests.INSTANCE = this;
+        this._client = client;
+    }
+
+    public enable(isEnabled:boolean) {
+        this._isEnabled = isEnabled;
+        if(this._isEnabled && !this._isInitialized) {
+            this._initialize();
+        }
+    }
+
+    public isInitialized() {
+        return this._isInitialized;
+    }
+
+    private _initialize() {
+        this._isInitialized = true;
+        this._initializeHttpModule(http);
+        this._initializeHttpModule(https);
+    }
+
+    private _initializeHttpModule(httpModule: any) {
+        const originalCreateServer = httpModule.createServer;
+        httpModule.createServer = (onRequest) => {
+            // todo: get a pointer to the server so the IP address can be read from server.address
+            return originalCreateServer((request: http.ServerRequest, response: http.ServerResponse) => {
+                if (this._isEnabled) {
+                    AutoCollectServerRequests.trackRequest(this._client, request, response);
+                }
+
+                if (typeof onRequest === "function") {
+                    onRequest(request, response);
+                }
+            });
+        };
+    }
+
+    /**
+     * Tracks a request synchronously (doesn't wait for response 'finish' event)
+     */
+    public static trackRequestSync(client: Client, request: http.ServerRequest, response:http.ServerResponse, ellapsedMilliseconds?: number, properties?:{ [key: string]: string; }, error?: any) {
+        if (!request || !response || !client) {
+            Logging.info("AutoCollectServerRequests.trackRequestSync was called with invalid parameters: ", !request, !response, !client);
+            return;
+        }
+
+        // store data about the request
+        var requestParser = new ServerRequestParser(request);
+
+        AutoCollectServerRequests.endRequest(client, requestParser, response, ellapsedMilliseconds, properties, error);
+    }
+
+    /**
+     * Tracks a request by listening to the response 'finish' event
+     */
+    public static trackRequest(client:Client, request:http.ServerRequest, response:http.ServerResponse, properties?:{ [key: string]: string; }) {
+        if (!request || !response || !client) {
+            Logging.info("AutoCollectServerRequests.trackRequest was called with invalid parameters: ", !request, !response, !client);
+            return;
+        }
+
+        // store data about the request
+        var requestParser = new ServerRequestParser(request);
+
+        // response listeners
+        if (response && response.once) {
+            response.once("finish", () => {
+                AutoCollectServerRequests.endRequest(client, requestParser, response, null, properties, null);
```

Comment:
I just noticed that in both the Server and Client auto collectors, it's currently not parsing or generating a hashed ikey and on the HTTP headers. This will be needed for stitching different AI components in the map. Perhaps that's the next step?